### PR TITLE
[WIP] Add 'split' support for ReadOnlySpan

### PR DESCRIPTION
**Pull request by grant-d**
Resolves https://github.com/dotnet/corefx/issues/26528

### CODE REVIEW

File: src/System.Private.CoreLib/shared/System/MemoryExtensions.Split.cs [hunk 1]
```diff
@@ -0,0 +1,105 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+namespace System
+{
+    public static partial class MemoryExtensions
+    {
+        public static SpanSplitEnumerator<T> Split<T>(this ReadOnlySpan<T> span, T separator, StringSplitOptions options = StringSplitOptions.None)
```
Comment (davidfowl) [inline src/System.Private.CoreLib/shared/System/MemoryExtensions.Split.cs]: StringSplitOptions is kinda weird.

Comment (grant-d) [reply src/System.Private.CoreLib/shared/System/MemoryExtensions.Split.cs]: Agree. That question came up in the design review, but it wasn't resolved.
https://github.com/dotnet/corefx/issues/26528#issuecomment-420629265

Comment (scalablecory) [reply src/System.Private.CoreLib/shared/System/MemoryExtensions.Split.cs]: The approved API does not have `StringSplitOptions`. It is in a comment here: https://github.com/dotnet/corefx/issues/26528#issuecomment-523152939

```c#
public static SpanSplitEnumerator<char> Split(this ReadOnlySpan<char> span);
public static SpanSplitEnumerator<char> Split(this ReadOnlySpan<char> span, char separator);
public static SpanSplitEnumerator<char> Split(this ReadOnlySpan<char> span, string separator);
```

Comment (grant-d) [reply src/System.Private.CoreLib/shared/System/MemoryExtensions.Split.cs]: Ah, I was confused, thanks.
What does the first overload do - it has no explicit `separator`?

**Comment by grant-d** [comment]
> Looks like the original issue top-level comment was not updated with approved API, so you're a tiny bit off here.
> 
> CC @terrajobst

Thanks, I was confused. I will realign.

File: src/System.Private.CoreLib/shared/System/MemoryExtensions.Split.cs (hunk 1 above)
Comment (scalablecory) [reply src/System.Private.CoreLib/shared/System/MemoryExtensions.Split.cs]: Splits on whitespace. I would follow the `string.Split(null)` rules and use `char.IsWhiteSpace`.

File: src/System.Private.CoreLib/shared/System/MemoryExtensions.Split.cs [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+                    ? _span
+                    : _span.Slice(_start);
+
+                int end = _start;
+                if (slice.Length > 0)
+                {
+                    int index;
+                    switch (_kind)
+                    {
+                        default: // TODO: Throw?
```
Comment (grant-d) [inline src/System.Private.CoreLib/shared/System/MemoryExtensions.Split.cs]: Question

File: src/System.Private.CoreLib/shared/System/MemoryExtensions.Split.cs [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+        public static SpanSplitEnumerator<char> Split(this ReadOnlySpan<char> span)
+            => new SpanSplitEnumerator<char>(span);
+
+        public static SpanSplitEnumerator<char> Split(this ReadOnlySpan<char> span, char separator)
+            => new SpanSplitEnumerator<char>(span, separator);
+
+        public static SpanSplitEnumerator<char> Split(this ReadOnlySpan<char> span, string separator)
+            => new SpanSplitEnumerator<char>(span, separator);
+
+        // TODO: Is this required (missed on the spec)?
```
Comment (grant-d) [inline src/System.Private.CoreLib/shared/System/MemoryExtensions.Split.cs]: Question

File: src/System.Private.CoreLib/shared/System/MemoryExtensions.Split.cs [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+        public static SpanSplitEnumerator<char> Split(this ReadOnlySpan<char> span, string separator, StringComparison comparisonType)
+            => new SpanSplitEnumerator<char>(span, separator, comparisonType);
```
Comment (grant-d) [inline src/System.Private.CoreLib/shared/System/MemoryExtensions.Split.cs]: Question

**Comment by grant-d** [comment]
I have aligned the spec and updated the implementation. I am not sure if brighter minds have an idea of how to support the single `SpanSplitEnumerator<T>` class with different implementations, but since it's a `ref struct` I could not use an `interface` or `inheritance`, and even composition was challenging since I needed a polymorphic state field regardless.

I opted for a `switch` but happy to change it if I am missing the blindingly obvious.

File: src/System.Private.CoreLib/shared/System/MemoryExtensions.Split.cs [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+
+                ReadOnlySpan<char> slice = _start == 0
+                    ? _span
+                    : _span.Slice(_start);
+
+                int end = _start;
+                if (slice.Length > 0)
+                {
+                    int index;
+                    switch (_kind)
```
Comment (grant-d) [inline src/System.Private.CoreLib/shared/System/MemoryExtensions.Split.cs]: See question in discussion below about this.

**Comment by GrabYourPitchforks** [comment]
Be aware that culture-aware searching could change the length of the matched strings. For example, "æ".StartsWith("ae") returns _true_ (even though the search term is longer than the source). And the string "sss" contains only _one_ Eszett character ('ß') - not two.

Depending on the `StringComparison` provided we'll need unit tests to make sure any optimizations we've added don't violate these special-cases.

**Comment by scalablecory** [comment]
> I have aligned the spec and updated the implementation. I am not sure if **brighter minds** have an idea of how to support the single `SpanSplitEnumerator<T>` class with different implementations

During API review we arrived at it being okay to use multiple return types as the expectation is that most people will use it with `foreach`.

That doesn't mean you *should*, of course. I would code a version with one type, and then a version using multiple types, and benchmark to see if there's a significant difference.

**Comment by scalablecory** [comment]
@grant-d are you planning to pick this up again?

**Comment by maryamariyan** [comment]
Thank you for your contribution. As announced in dotnet/coreclr#27549 this repository will be moving to dotnet/runtime on November 13. If you would like to continue working on this PR after this date, the easiest way to move the change to dotnet/runtime is:
1.	In your coreclr repository clone, create patch by running `git format-patch origin`
2.	In your runtime repository clone, apply the patch by running `git apply --directory src/coreclr <path to the patch created in step 1>`

**Comment by grant-d** [comment]
> @grant-d are you planning to pick this up again?

I was hoping to but I am just not getting to it. Can you un-assign me on the related issue please, https://github.com/dotnet/corefx/issues/26528

Thanks

**Pull request closed without merging**